src/Albany_Application.cpp

```diff
@@ -395,6 +395,45 @@ Application::initialSetUp(const RCP<Teuchos::ParameterList>& params)
 #endif
   }
 
+  //validate Hessian parameters
+  if(problemParams->isSublist("Hessian")) {
+    Teuchos::RCP<Teuchos::ParameterList> validHessianParams = Teuchos::rcp(new Teuchos::ParameterList("validHessianParams"));
+    validHessianParams->set<bool>("Write Hessian MatrixMarket", false);
+    validHessianParams->set<bool>("Use AD for Hessian-vector products (default)", true);
+
+    validHessianParams->sublist("Residual").set<bool>("Use AD for Hessian-vector products (default)", false);
+    validHessianParams->sublist("Residual").set<std::string>("Enable AD for Hessian-vector product contributions of", "(x,x) (p0,p0) (p0,p1) (p1,p0) (p1,p1)");
+    validHessianParams->sublist("Residual").set<std::string>("Disable AD for Hessian-vector product contributions of", "(x,p0) (x,p1) (p0,x) (p1,x)");
+
+    // We want to allow the user to add hessian settings for responses (parameters) even if those responses (parameters) are not used.
+    int maxNumResonses(3),maxNumParameters(3);
+
+    //Make sure that the problem does not require more responses (parameters) than anticipated. 
+    if (problemParams->isSublist("Response Functions")) 
+      maxNumResonses = std::max(maxNumResonses, problemParams->sublist("Response Functions").get<int>("Number Of Responses"));
+    if (problemParams->isSublist("Parameters")) 
+      maxNumParameters = std::max(maxNumParameters, problemParams->sublist("Parameters").get<int>("Number Of Parameters"));
+
+    for(int response_index = 0;  response_index < maxNumResonses; response_index++) {
+      auto& validHessianResponseParams = validHessianParams->sublist(util::strint("Response", response_index));
+      validHessianResponseParams.set<bool>("Use AD for Hessian-vector products (default)", false);
+      validHessianResponseParams.set<bool>("Reconstruct H_pp", false);
+      validHessianResponseParams.set<std::string>("Enable AD for Hessian-vector product contributions of", "(x,x) (p0,p0) (p0,p1) (p1,p0) (p1,p1)");
```

Comment:
Can I ask what p0 and p1 refer to? Are you assuming there are only two parameters? Or is this to say that the two ps in H_pp could refer to different parameters?

Reply:
@bartgol it is the later. It suggests that $\boldsymbol{p}^{(1)} != \boldsymbol{p}^{(2)}$.

Reply:
Yes, in this case we are considering two parameters, but you could have terms like   (p0, p4), (x, p17), etc.